File: java/token/token.go
package token

type Token int

const (
	ILLEGAL Token = iota
	EOF
	META

	// literals
	literalBegin
	IDENT
	BOOL
	CHAR
	INT
	FLOAT
	STRING
	NULL
	literalEnd

	// keywords
	keywordBegin
	Break
	Case
	Const
	Continue
	Default
	Else
	Enum
	For
	Function
	If
	Import
	Namespace
	Pointer
	Public
	Return
	Sizeof
	Struct
	Switch
	This
	Var
	keywordEnd

	// scalars
	scalarBegin
	Bool
	numberBegin
	integerBegin
	Int8
	Int16
	Int32
	Int64
	Uint8
	Uint16
	Uint32
	Uint64
	integerEnd
	floatBegin
	Float16
	Float32
	Float64
	floatEnd
	numberEnd
	Void
	scalarEnd

	// operators
	operatorBegin
	LeftParen
	RightParen
	LeftBracket
	RightBracket
	LeftBrace
	RightBrace

	Plus
	Minus
	Mul
	Div
	Rem
	BitAnd
	BitOr
	BitXor
	Complement
	Not
	Less
	Greater

	assignBegin
	Assign
	PlusAssign
	MinusAssign
	MulAssign
	DivAssign
	RemAssign
	XorAssign
	AndAssign
	OrAssign
	LeftShift
	RightShift
	LeftShiftAssign
	RightShiftAssign
	assignEnd

	Equal
	NotEqual
	LessEqual
	GreaterEqual
	And
	Or
	PlusPlus
	MinusMinus

	Comma
	Colon
	Semi
	Dot
	operatorEnd
)

var (
	tokenStrings = [...]string{
		IDENT:  "identifier",
		BOOL:   "bool_literal",
		CHAR:   "char_literal",
		INT:    "int_literal",
		FLOAT:  "float_literal",
		STRING: "string_literal",
		NULL:   "null",

		Break:     "break",
		Case:      "case",
		Const:     "const",
		Continue:  "continue",
		Default:   "default",
		Else:      "else",
		Enum:      "enum",
		For:       "for",
		Function:  "function",
		If:        "if",
		Import:    "import",
		Namespace: "namespace",
		Pointer:   "pointer",
		Public:    "public",
		Return:    "return",
		Sizeof:    "sizeof",
		Struct:    "struct",
		Switch:    "switch",
		This:      "this",
		Var:       "var",

		Bool:    "bool",
		Int8:    "i8",
		Int16:   "i16",
		Int32:   "i32",
		Int64:   "i64",
		Uint8:   "u8",
		Uint16:  "u16",
		Uint32:  "u32",
		Uint64:  "u64",
		Float16: "f16",
		Float32: "f32",
		Float64: "f64",
		Void:    "void",

		LeftParen:    "(",
		RightParen:   ")",
		LeftBracket:  "[",
		RightBracket: "]",
		LeftBrace:    "{",
		RightBrace:   "}",

		Plus:             "+",
		Minus:            "-",
		Mul:              "*",
		Div:              "/",
		Rem:              "%",
		BitAnd:           "&",
		BitOr:            "|",
		BitXor:           "^",
		Complement:       "~",
		Not:              "!",
		Assign:           "=",
		Less:             "<",
		Greater:          ">",
		PlusAssign:       "+=",
		MinusAssign:      "-=",
		MulAssign:        "*=",
		DivAssign:        "/=",
		RemAssign:        "%=",
		XorAssign:        "^=",
		AndAssign:        "&=",
		OrAssign:         "|=",
		LeftShift:        "<<",
		RightShift:       ">>",
		LeftShiftAssign:  "<<=",
		RightShiftAssign: ">>=",
		Equal:            "==",
		NotEqual:         "!=",
		LessEqual:        "<=",
		GreaterEqual:     ">=",
		And:              "&&",
		Or:               "||",
		PlusPlus:         "++",
		MinusMinus:       "--",
		Comma:            ",",
		Colon:            ":",
		Semi:             ";",
		Dot:              ".",
	}

	tokens map[string]Token
)

func init() {
	tokens = make(map[string]Token)

	for i := keywordBegin + 1; i < keywordEnd; i++ {
		tokens[tokenStrings[i]] = i
	}

	for i := scalarBegin + 1; i < scalarEnd; i++ {
		tokens[tokenStrings[i]] = i
	}

	for i := operatorBegin + 1; i < operatorEnd; i++ {
		tokens[tokenStrings[i]] = i
	}

	operatorRoot = &operatorNode{
		children: make(map[byte]*operatorNode),
		token:    ILLEGAL,
	}

	for i := operatorBegin + 1; i < operatorEnd; i++ {
		operatorRoot.insert(tokenStrings[i])
	}
}

func ReadToken(literal string) Token {
	if token, ok := tokens[literal]; ok {
		return token
	}
	if literal == "true" || literal == "false" {
		return BOOL
	}
	if literal == "null" {
		return NULL
	}
	return IDENT
}

func (t Token) String() string {
	if 0 <= t && t < Token(len(tokenStrings)) {
		return tokenStrings[t]
	}
	return ""
}

func (t Token) IsLiteral() bool {
	return literalBegin < t && t < literalEnd
}

func (t Token) IsOperator() bool {
	return operatorBegin < t && t < operatorEnd
}

func (t Token) IsKeyword() bool {
	return keywordBegin < t && t < keywordEnd
}

func (t Token) IsScalar() bool {
	return scalarBegin < t && t < scalarEnd
}

func (t Token) IsInteger() bool {
	return integerBegin < t && t < integerEnd
}

func (t Token) IsFloat() bool {
	return floatBegin < t && t < floatEnd
}

func (t Token) IsNumber() bool {
	return numberBegin < t && t < numberEnd
}

func (t Token) IsAssign() bool {
	return assignBegin < t && t < assignEnd
}

func (t Token) Precedence() int {
	switch t {
	case Assign, MulAssign, DivAssign, RemAssign, PlusAssign, MinusAssign,
		LeftShiftAssign, RightShiftAssign, AndAssign, OrAssign, XorAssign:
		return 1

	case Or:
		return 2

	case And:
		return 3

	case BitOr:
		return 4

	case BitXor:
		return 5

	case BitAnd:
		return 6

	case Equal, NotEqual:
		return 7

	case Less, LessEqual, Greater, GreaterEqual:
		return 8

	case LeftShift, RightShift:
		return 9

	case Plus, Minus:
		return 10

	case Mul, Div, Rem:
		return 11
	}
	return 0
}

File: MicroPanda/Token/Token.cs
namespace MicroPanda.Token;

using System.Collections.Generic;
using System.Text;

internal enum Token
{
	ILLEGAL,
	EOF,
	COMMENT,
	ANNOTATION,
	NEWLINE,

	// literals
	_literalBegin,
	IDENT,
	BOOL,
	CHAR,
	INT,
	FLOAT,
	STRING,
	NULL,
	_literalEnd,

	// keywords
	_keywordBegin,
	Break,
	Case,
	Const,
	Continue,
	Default,
	Else,
	Enum,
	For,
	Function,
	If,
	Namespace,
	Pointer,
	Public,
	Return,
	Sizeof,
	Struct,
	Switch,
	This,
	Using,
	Var,
	_keywordEnd,

	// scalar
	_scalarBegin,
	Bool,
	_numberBegin,
	_integerBegin,
	Int8,
	Int16,
	Int32,
	Int64,
	Uint8,
	Uint16,
	Uint32,
	Uint64,
	_integerEnd,
	_floatBegin,
	integerEnd,
	floatBegin,
	Float16,
	Float32,
	Float64,
	_floatEnd,
	_numberEnd,
	Void,
	_scalarEnd,

	// operators
	_operatorBegin,
	LeftParen,
	RightParen,
	LeftBracket,
	RightBracket,
	LeftBrace,
	RightBrace,

	Plus,
	Minus,
	Mul,
	Div,
	Less,
	Greater,
	Rem,
	BitAnd,
	BitOr,
	BitXor,
	Complement,
	Not,
	LeftShift,
	RightShift,

	_assignBegin,
	Assign,
	PlusAssign,
	MinusAssign,
	MulAssign,
	DivAssign,
	RemAssign,
	XorAssign,
	AndAssign,
	OrAssign,
	LeftShiftAssign,
	RightShiftAssign,
	_assignEnd,

	Equal,
	NotEqual,
	LessEqual,
	GreaterEqual,
	And,
	Or,
	PlusPlus,
	MinusMinus,

	Comma,
	Colon,
	Semi,
	Dot,
	_operatorEnd
}

internal static class TokenHelper
{
	internal static bool IsLiteral(Token token) => Token._literalBegin < token && token < Token._literalEnd;

	internal static bool IsOperator(Token token) => Token._operatorBegin < token && token < Token._operatorEnd;

	internal static bool IsKeyword(Token token) => Token._keywordBegin < token && token < Token._keywordEnd;

	internal static bool IsScalar(Token token) => Token._scalarBegin < token && token < Token._scalarEnd;

	internal static bool IsInteger(Token token) => Token._integerBegin < token && token < Token._integerEnd;

	internal static bool IsFloat(Token token) => Token._floatBegin < token && token < Token._floatEnd;

	internal static bool IsNumber(Token token) => Token._numberBegin < token && token < Token._numberEnd;

	internal static bool IsAssign(Token token) => Token._assignBegin < token && token < Token._assignEnd;

	internal static bool IsOperator(string literal) => _operations.Contains(literal);

	internal static Token FromString(string literal)
	{
		if (_string2Token.TryGetValue(literal, out var token))
		{
			return token;
		}
		else if (literal == "true" || literal == "false")
		{
			return Token.BOOL;
		}
		else if (literal == "null")
		{
			return Token.NULL;
		}
		return Token.IDENT;
	}

	internal static string ToString(Token token)
	{
		return _token2String[token];
	}

	internal static int Precedence(Token token)
	{
        return token switch
        {
            Token.Assign or Token.MulAssign or Token.DivAssign or Token.RemAssign or Token.PlusAssign or Token.MinusAssign or Token.LeftShiftAssign or Token.RightShiftAssign or Token.AndAssign or Token.OrAssign or Token.XorAssign => 1,
            Token.Or => 2,
            Token.And => 3,
            Token.BitOr => 4,
            Token.BitXor => 5,
            Token.BitAnd => 6,
            Token.Equal or Token.NotEqual => 7,
            Token.Less or Token.LessEqual or Token.Greater or Token.GreaterEqual => 8,
            Token.LeftShift or Token.RightShift => 9,
            Token.Plus or Token.Minus => 10,
            Token.Mul or Token.Div or Token.Rem => 11,
            _ => 0,
        };
    }

	static TokenHelper()
	{
		foreach (var pair in _token2String)
		{
			_string2Token[pair.Value] = pair.Key;
		}

		for (var i = Token._operatorBegin + 1; i < Token._operatorEnd; i++)
		{
			if (i != Token._assignBegin && i != Token._assignEnd)
			{
				_operations.Add(_token2String[i]);
			}
		}
	}

	private static readonly Dictionary<Token, string> _token2String = new()
	{
		{ Token.ILLEGAL, "ILLEGAL" },
		{ Token.EOF, "EOF" },
		{ Token.COMMENT, "COMMENT" },
		{ Token.ANNOTATION, "ANNOTATION" },
		{ Token.NEWLINE, "NEWLINE" },

		{ Token.IDENT, "identifier" },
		{ Token.BOOL, "bool_literal" },
		{ Token.CHAR, "char_literal" },
		{ Token.INT, "int_literal" },
		{ Token.FLOAT, "float_literal" },
		{ Token.STRING, "string_literal" },
		{ Token.NULL, "null" },

		{ Token.Break, "break" },
		{ Token.Case, "case" },
		{ Token.Const, "const" },
		{ Token.Continue, "continue" },
		{ Token.Default, "default" },
		{ Token.Else, "else" },
		{ Token.Enum, "enum" },
		{ Token.For, "for" },
		{ Token.Function, "function" },
		{ Token.If, "if" },
		{ Token.Namespace, "namespace" },
		{ Token.Pointer, "pointer" },
		{ Token.Public, "public" },
		{ Token.Return, "return" },
		{ Token.Sizeof, "sizeof" },
		{ Token.Struct, "struct" },
		{ Token.Switch, "switch" },
		{ Token.This, "this" },
		{ Token.Using, "using" },
		{ Token.Var, "var" },

		{ Token.Bool, "bool" },
		{ Token.Int8, "i8" },
		{ Token.Int16, "i16" },
		{ Token.Int32, "i32" },
		{ Token.Int64, "i64" },
		{ Token.Uint8, "u8" },
		{ Token.Uint16, "u16" },
		{ Token.Uint32, "u32" },
		{ Token.Uint64, "u64" },
		{ Token.Float16, "f16" },
		{ Token.Float32, "f32" },
		{ Token.Float64, "f64" },
		{ Token.Void, "void" },

		{ Token.LeftParen, "(" },
		{ Token.RightParen, ")" },
		{ Token.LeftBracket, "[" },
		{ Token.RightBracket, "]" },
		{ Token.LeftBrace, "{" },
		{ Token.RightBrace, "}" },

		{ Token.Plus, "+" },
		{ Token.Minus, "-" },
		{ Token.Mul, "*" },
		{ Token.Div, "/" },
		{ Token.Less, "<" },
		{ Token.Greater, ">" },
		{ Token.Rem, "%" },
		{ Token.BitAnd, "&" },
		{ Token.BitOr, "|" },
		{ Token.BitXor, "^" },
		{ Token.Complement, "~" },
		{ Token.Not, "!" },
		{ Token.LeftShift, "<<" },
		{ Token.RightShift, ">>" },

		{ Token.Assign, "=" },
		{ Token.PlusAssign, "+=" },
		{ Token.MinusAssign, "-=" },
		{ Token.MulAssign, "*=" },
		{ Token.DivAssign, "/=" },
		{ Token.RemAssign, "%=" },
		{ Token.XorAssign, "^=" },
		{ Token.AndAssign, "&=" },
		{ Token.OrAssign, "|=" },
		{ Token.LeftShiftAssign, "<<=" },
		{ Token.RightShiftAssign, ">>=" },

		{ Token.Equal, "==" },
		{ Token.NotEqual, "!=" },
		{ Token.LessEqual, "<=" },
		{ Token.GreaterEqual, ">=" },
		{ Token.And, "&&" },
		{ Token.Or, "||" },
		{ Token.PlusPlus, "++" },
		{ Token.MinusMinus, "--" },

		{ Token.Comma, "," },
		{ Token.Colon, ":" },
		{ Token.Semi, ";" },
		{ Token.Dot, "." }
	};

	private static readonly Dictionary<string, Token> _string2Token = [];

	private static readonly HashSet<string> _operations = [];
}